cdap-ui/app/features/etlapps/services/etl-plugin-config-factory.js

```diff
@@ -0,0 +1,24 @@
+angular.module(PKG.name + '.feature.etlapps')
+  .service('PluginConfigFactory', function(MyDataSource, $q) {
+    this.plugins = {};
+
+    this.fetch = function(scope, templateid, pluginid) {
+      var dataSrc = new MyDataSource(scope);
+      var defer = $q.defer();
+
+      if (this.plugins[templateid+pluginid]) {
+        return $q.when(this.plugins[templateid+pluginid]);
+      }
+
+      dataSrc.config({
+        templateid: templateid, //'etlRealtime',
+        pluginid: pluginid //'TwitterSource'
+      })
+        .then(function(res) {
+          this.plugins[templateid+pluginid] = res;
+          defer.resolve(this.plugins[templateid+pluginid]);
+        }.bind(this));
+      return defer.promise;
```

Comment:
If you are already returning a promise, do you need to bind(this) as well to the 'then' function?